### DISCUSSION TITLE
[A-L] (14 marzo 2024) Esercizio 6

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#22] [Esercitazioni A-L]
Descrivere in pseudo-codice un algoritmo che, dato un grafo **non diretto** $G$, descrivere un algoritmo che ne orienta gli archi in modo da creare un grafo $G'$ diretto e aciclico. Questo algoritmo deve avere tempo di esecuzione $\theta(|V|+|E|)$.

**Comment by daromacarlo** [comment]
```
def ord_and_acyclic (G: non-directed) => list, list
    V(G') = V(G)         //nodes in G'
    E(G') = []           //arcs in G'
    stack = {}           //initialize an empty stack
    vis = [0,...,0]      //O(N)time complexity
    x = random(V(G))
    while stack is not empty:
        y = stack.top()
	if y.adjacent != []:
	    z = y.adjacent[0]
	    if vis[z] == 0:
	    
	        stack.push(z)
	        E(G').append((y,z))
                z.adjacent.remove(0)

	   elif vis[z] == 1 && z is not father:
	   
		E(G').append((z,y))
                z.adjacent.remove(0)
	   
	  father = y
	  else:
		stack.pop()
    return V(G'),E(G')
```
- L'idea è quella di percorrere il grafo facendo diventare gli archi su cui passo degli archi entranti al nodo corrente, quando trovo un nodo già visitato per evitare di formare un ciclo ordinato, l'arco che collega il nodo corrente al nuovo nodo trovato sarà un arco uscente dal nodo corrente.
Non sono sicuro di qusta soluzione, volevo scrivere un algoritmo che sfruttasse l'albero di visita e i suoi gli archi all'indietro.
Inoltre non sono sicuro che questo algoritmo restituisca sempre dei grafi aciclici.

- il costo compuztazionale è $O(M + N)$ perché il ciclo while viene ripetuto $O(N)$ volte e ogni ciclo ha costo $O(deg(v))$, quindi il ciclo while ha costo $O(\sum_{v \in V(G)}deg(v))$, mentre la costruzione della lista "vis" ha costo $O(N)$.

**Reply by Frisbon** [reply]
Mi sembra funzionare perfettamente, sugli esempi che faccio non vengono creati cicli
Complimenti per la soluzione

**Reply by daromacarlo** [reply]
Grazie 😊

**Reply by daromacarlo** [reply]
Ho rifatto l'esercizio sfruttando l'arborescenza e gli archi all'indietro
```c
/*
 l'idea è quella di esplorare tutti gli archi e dare loro un verso, visitare il nuovo grafo
 ordinato, trovare una arborescenza, classificare gli archi non appartenenti all'arborescenza e 
 invertire gli archi all'indietro per renderli in avanti.
*/
//----
def esplora_e_assegna_verso(G, v, insieme_archi_ordinati)
	for w in v.adiacenti(){
		if (v,w) or (w,v) not in insieme_archi_ordinati{
			insieme_archi_ordinati += [(v,w)]
			esplora_e_assegna_verso(G, w, insieme_archi_ordinati)
		}
	}

// costo = O(m)

//----
def dfs_con_contatore(G, v, arb, vis, CC, counter)
	vis[v] = 1
	v.CC[t] = counter
	counter ++
	for w in v.uscenti{
		if vis[w] == 0{
			dfs_con_contatore(G, w, arb, vis, CC, counter)
			arb.append(v,w)
		}
	}
	v.CC[T] = counter

// costo = O(n + m)

//----
def trova_e_inverti_archi_indietro(G, arb, CC)
	for e = (x,y) in E(G) and not in arb{
		if x.CC[t] > y.CC[t] && x.CC[T] < y.CC[T]{		         //trovo un arco all'indietro
			E(G).remove(x,y)
			E(G).add(y,x)							 // cambio verso
		}
	}

// costo = O(m)

//----
def main(G = (V(G), E(G)) )  , v = random(V(G)))
	esplora_e_assegna_verso(G, v, insieme_archi_ordinati = [])
	G' = (V(G), insieme_archi_ordinati)
	arb = []							//verrà modificato dalla prossima funzione
	CC = []								//verrà modificato dalla prossima funzione
	dfs_con_contatore(G', v, arb, vis = [0,...,0], CC, counter = 0)
	trova_e_inverti_archi_indietro(G', arb, CC)
	return G'
```

**Reply by daromacarlo** [reply]
Rifatto più ez:
![immagine](https://github.com/user-attachments/assets/bb2970a8-86b0-4574-9740-ebac38e01d13)
![immagine](https://github.com/user-attachments/assets/a084b0a1-58db-40c9-9fb6-37c1ac3564a8)

**Comment by LeoCosta02** [comment]
```C
//Input: un grafo non diretto G=(V,E)
//Output: un grafo diretto e aciclico G'

main_dag(G){
vis[n]: array = [0...0]
t[n]: array=[0..0]
int cont=0
  for each u di G{
   Arc[u] // lista di adiacenza del nodo u vuota 
  }
  
    for each u di G{
     if vis[u]==0
	 Dfs_visit(G,u,u,t,cont)
	}
}


Dfs_visit(G,u,z,t,cont){
 vis[u]=1
 cont++
 t[u]=cont
  for each v adiacente di u {
   if vis[v]==0{
   Arc[u].append(v)
   Dfs_visit(G,v,u,t,cont)
   }
     else{
		 if(v != z AND t[u]>t[v])
		  Arc[v].append(u) //caso in cui ho arco all indietro 
       else{
         if(t[u]<t[v])
        return // arco in avanti
       }
        
	  }
```

**Reply by LeoCosta02** [reply]
Ho ripreso la soluzione del professore, l'unica cosa che ho aggiunto è il parametro del nodo padre che viene passata nella DFS altrimenti nell'else vengono considerati anche i casi in cui da un nodo ritorno indietro al suo padre e viene aggiunto nella lista di adiacenza due volte lo stesso nodo

**Reply by IonutCicio** [reply]
Quel $z$ non basta, consideriamo questo esempio con 3 vertici

![image](https://github.com/sapienzastudentsnetwork/progalg2324/assets/40034429/cdf2f327-0acf-475f-86cb-acbf523de500)

La visita parte dal vertice 1, gli archi dell'arborescenza sono quelli in giallo, gli archi del nuovo grafo sono quelli in rosso, e l'arco verde è problematico... ma perché?

Perché quando arrivo a 3, vedo che il vertice 1 l'ho già visitato, e aggiungo agli archi di 1 un arco che va verso 3 _(ovvero un arco in avanti, così non ho cicli)_

Il problema è che quando nel dfs torno indietro, e da 1 provo a visitare 3, vedo che 3 l'ho già visitato, quindi aggiungo un agli archi di 3 un arco che va verso 1..., che diventa un arco all'indietro! _(che non vogliamo, perché crea un ciclo!)_

**Reply by LeoCosta02** [reply]
Ho aggiornato il codice, adesso dovrebbe andare anche per il caso mostrato da @CuriousCI

**Comment by IonutCicio** [comment]
```Rust
pub fn to_dag(graph: &[Vec<usize>]) -> Vec<Vec<usize>> {
    let mut new_graph = vec![vec![]; graph.len()];
    let mut first_visit = vec![0; graph.len()];
    let mut visit = 1;

    dag_dfs(graph, 0, 0, &mut visit, &mut first_visit, &mut new_graph);

    new_graph
}

pub fn dag_dfs(
    graph: &[Vec<usize>],
    x: usize,
    z: usize,
    visit: &mut usize,
    first_visit: &mut Vec<usize>,
    new_graph: &mut [Vec<usize>],
) {
    first_visit[x] = *visit;
    *visit += 1;

    for &y in &graph[x] {
        if first_visit[y] == 0 {
            new_graph[x].push(y);
            dag_dfs(graph, y, x, visit, first_visit, new_graph)
        } else if first_visit[x] > first_visit[y] && y != z {
            new_graph[y].push(x);
        }
    }
}
```
